File: variable-streams/c23/src/bit_stream_writer.c
#include "bit_stream.h"

// Helper function to create a BitStreamResult with an error
static BitStreamResult create_error_result(BitStreamErrorCode code, int io_errno) {
    BitStreamResult result;
    result.success = false;
    result.error.code = code;
    result.error.io_errno = io_errno;
    return result;
}

// Helper function to create a BitStreamResult with success status
static BitStreamResult create_success_result(void) {
    BitStreamResult result;
    result.success = true;
    result.error.code = BIT_STREAM_ERROR_NONE;
    result.error.io_errno = 0;
    return result;
}

BitStreamWriter* bit_stream_writer_new(FILE* file) {
    return bit_stream_writer_with_capacity(file, 4096);
}

BitStreamWriter* bit_stream_writer_with_capacity(FILE* file, size_t capacity) {
    BitStreamWriter* writer = (BitStreamWriter*)malloc(sizeof(BitStreamWriter));
    if (writer == NULL) {
        return NULL;
    }
    
    writer->file = file;
    writer->buffer = (uint8_t*)malloc(capacity);
    if (writer->buffer == NULL) {
        free(writer);
        return NULL;
    }
    
    writer->buffer_capacity = capacity;
    writer->byte_pos = 0;
    writer->bit_pos = 0;
    
    // Initialize buffer to zeros
    memset(writer->buffer, 0, capacity);
    
    return writer;
}

void bit_stream_writer_free(BitStreamWriter* writer) {
    if (writer != NULL) {
        free(writer->buffer);
        free(writer);
    }
}

// Flushes the internal buffer to the underlying file
static BitStreamResult flush_buffer(BitStreamWriter* writer) {
    if (writer->byte_pos > 0) {
        // Write the buffer to the underlying file
        size_t bytes_written = fwrite(writer->buffer, 1, writer->byte_pos, writer->file);
        
        if (bytes_written != writer->byte_pos) {
            return create_error_result(BIT_STREAM_ERROR_IO, errno);
        }
        
        // Reset buffer position
        writer->byte_pos = 0;
        
        // Clear the first byte for the next write
        writer->buffer[0] = 0;
    }
    
    return create_success_result();
}

BitStreamResult bit_stream_writer_write_bits(BitStreamWriter* writer, uint64_t value, uint8_t bit_count) {
    if (bit_count == 0 || bit_count > 64) {
        return create_error_result(BIT_STREAM_ERROR_INVALID_BIT_COUNT, 0);
    }
    
    uint8_t bits_written = 0;
    
    // Ensure the buffer has enough space
    if (writer->byte_pos >= writer->buffer_capacity) {
        BitStreamResult flush_result = flush_buffer(writer);
        if (!flush_result.success) {
            return flush_result;
        }
    }
    
    while (bits_written < bit_count) {
        uint8_t bits_left_in_byte = 8 - writer->bit_pos;
        uint8_t bits_to_write = (bit_count - bits_written < bits_left_in_byte) ? 
                                (bit_count - bits_written) : bits_left_in_byte;
        
        // Extract the bits to write (LSB first)
        uint8_t shift = bits_written;
        uint64_t mask = ((1ULL << bits_to_write) - 1) << shift;
        uint8_t extracted_bits = (uint8_t)((value & mask) >> shift);
        
        // Write the bits to the current byte
        uint8_t byte_shift = bits_left_in_byte - bits_to_write;
        uint8_t byte_mask = ~(((1U << bits_to_write) - 1) << byte_shift);
        writer->buffer[writer->byte_pos] &= byte_mask;
        writer->buffer[writer->byte_pos] |= extracted_bits << byte_shift;
        
        // Update position
        bits_written += bits_to_write;
        writer->bit_pos += bits_to_write;
        if (writer->bit_pos == 8) {
            writer->byte_pos += 1;
            writer->bit_pos = 0;
            
            // If the buffer is full, flush it
            if (writer->byte_pos >= writer->buffer_capacity) {
                BitStreamResult flush_result = flush_buffer(writer);
                if (!flush_result.success) {
                    return flush_result;
                }
            } else {
                // Initialize the next byte to zero
                writer->buffer[writer->byte_pos] = 0;
            }
        }
    }
    
    return create_success_result();
}

BitStreamResult bit_stream_writer_write_bits_u128(BitStreamWriter* writer, UInt128 value, uint8_t bit_count) {
    if (bit_count == 0 || bit_count > 128) {
        return create_error_result(BIT_STREAM_ERROR_INVALID_BIT_COUNT, 0);
    }
    // For bit counts up to 64, we can use the existing write_bits method
    if (bit_count <= 64) {
        return bit_stream_writer_write_bits(writer, value.low, bit_count);
    }
    // For bit counts > 64, write the low 64 bits first (LSB), then the high bits
    BitStreamResult low_result = bit_stream_writer_write_bits(writer, value.low, 64);
    if (!low_result.success) {
        return low_result;
    }
    return bit_stream_writer_write_bits(writer, value.high, bit_count - 64);
}

BitStreamResult bit_stream_writer_write_bit_value(BitStreamWriter* writer, BitValue value, uint8_t bit_count) {
    uint8_t actual_bit_count = bit_count;
    if (actual_bit_count == 0) {
        actual_bit_count = bit_value_bit_count(&value);
    }
    
    if (actual_bit_count <= 64) {
        return bit_stream_writer_write_bits(writer, bit_value_to_u64(&value), actual_bit_count);
    } else {
        return bit_stream_writer_write_bits_u128(writer, bit_value_to_u128(&value), actual_bit_count);
    }
}

BitStreamResult bit_stream_writer_flush(BitStreamWriter* writer) {
    // If there are any bits in the current byte, write it
    if (writer->bit_pos > 0) {
        writer->byte_pos += 1;
        writer->bit_pos = 0;
    }
    
    // Flush the buffer
    BitStreamResult flush_result = flush_buffer(writer);
    if (!flush_result.success) {
        return flush_result;
    }
    
    // Flush the underlying file
    if (fflush(writer->file) != 0) {
        return create_error_result(BIT_STREAM_ERROR_IO, errno);
    }
    
    return create_success_result();
}

File: variable-streams/python/src/variable_streams/bit_stream_writer.py
"""BitStreamWriter class for writing bits to an external stream."""

from typing import BinaryIO, Optional, Union
from .bit_value import BitValue, InvalidBitCountError


class BitStreamWriter:
    """A writer that allows writing individual bits to an underlying stream.
    
    This class provides methods for writing bits to any binary stream that
    supports the write method.
    """
    
    def __init__(self, stream: BinaryIO, buffer_size: int = 4096):
        """Initialize a new BitStreamWriter.
        
        Args:
            stream: The binary stream to write to.
            buffer_size: The size of the internal buffer in bytes.
        """
        self._stream = stream
        self._buffer = bytearray(buffer_size)
        self._byte_pos = 0
        self._bit_pos = 0
        self._buffer_size = buffer_size
    
    def _flush_buffer(self) -> None:
        """Flush the internal buffer to the underlying stream."""
        if self._byte_pos > 0:
            # Write the buffer to the underlying stream
            self._stream.write(self._buffer[:self._byte_pos])
            self._byte_pos = 0
            self._buffer[0] = 0  # Clear the first byte for the next write
    
    def flush(self) -> None:
        """Flush any remaining bits to the underlying stream."""
        # If there are any bits in the current byte, write it
        if self._bit_pos > 0:
            self._byte_pos += 1
            self._bit_pos = 0
        
        # Flush the buffer
        self._flush_buffer()
        
        # Flush the underlying stream
        self._stream.flush()
    
    def write_bits(self, value: int, bit_count: int) -> None:
        """Write up to 64 bits to the stream in LSB order (least significant bit first)."""
        if bit_count <= 0 or bit_count > 64:
            raise InvalidBitCountError(f"Bit count must be between 1 and 64, got {bit_count}")
        
        bits_written = 0
        
        # Ensure the buffer has enough space
        if self._byte_pos >= self._buffer_size:
            self._flush_buffer()
        
        while bits_written < bit_count:
            bits_left_in_byte = 8 - self._bit_pos
            bits_to_write = min(bit_count - bits_written, bits_left_in_byte)
            
            # Extract the bits to write (LSB first)
            shift = bits_written
            mask = ((1 << bits_to_write) - 1) << shift
            extracted_bits = (value & mask) >> shift
            
            # Write the bits to the current byte
            byte_shift = bits_left_in_byte - bits_to_write
            byte_mask = ~(((1 << bits_to_write) - 1) << byte_shift) & 0xFF
            self._buffer[self._byte_pos] &= byte_mask
            self._buffer[self._byte_pos] |= (extracted_bits << byte_shift)
            
            # Update position
            bits_written += bits_to_write
            self._bit_pos += bits_to_write
            if self._bit_pos == 8:
                self._byte_pos += 1
                self._bit_pos = 0
                
                # If the buffer is full, flush it
                if self._byte_pos >= self._buffer_size:
                    self._flush_buffer()
    
    def write_bits_u128(self, value: int, bit_count: int) -> None:
        """Write up to 128 bits to the stream in LSB order (lower bits first)."""
        if bit_count <= 0 or bit_count > 128:
            raise InvalidBitCountError(f"Bit count must be between 1 and 128, got {bit_count}")
        if bit_count <= 64:
            return self.write_bits(value, bit_count)
        # For bit counts > 64, write the lower 64 bits first (LSB), then the higher bits
        low_value = value & ((1 << 64) - 1)
        high_bits = bit_count - 64
        high_value = value >> 64
        self.write_bits(low_value, 64)
        self.write_bits(high_value, high_bits)
    
    def write_bit_value(self, value: BitValue, bit_count: Optional[int] = None) -> None:
        """Write a BitValue to the stream.
        
        Args:
            value: The BitValue to write.
            bit_count: The number of bits to write (1-128). If None, uses the bit count of the BitValue.
            
        Raises:
            InvalidBitCountError: If bit_count is invalid (0 or > 128).
        """
        bit_count = bit_count if bit_count is not None else value.bit_count
        
        if bit_count <= 64:
            self.write_bits(int(value), bit_count)
        else:
            self.write_bits_u128(int(value), bit_count)
    
    def __del__(self) -> None:
        """Ensure the buffer is flushed when the object is garbage collected."""
        try:
            self.flush()
        except:
            # Ignore errors during garbage collection
            pass